### ISSUE TITLE
Documentation for new PSF code

### ISSUE THREAD

**Issue by robelgeda** [spacetelescope/STScI-STIPS#81]
Detailed docs for #77

**Comment by benw1** [comment]
How do I specify a gridded PSF?  Is there a keyword to set?  I have not been able to test this feature since I can only get it to produce the default single PSF model.

**Comment by benw1** [comment]
Also, it appears to be convolving over a grid, even when only a single PSF is being used:
2020-07-03 13:01:19,701 INFO: Convolving SCA01 with PSF grid
2020-07-03 13:01:19,705 INFO: PSF Shape: [4100 4100]; Current Shape: [45060 45060]
2020-07-03 13:01:19,705 INFO: Choosing between 8191-4100=4091 and 4100+45060-1=49159
2020-07-03 13:01:19,705 INFO: Using overlapping arrays of size (4091, 4091)
2020-07-03 13:01:19,705 INFO: SCA01: Starting Convolution at Fri Jul  3 13:01:19 2020
2020-07-03 13:01:19,706 INFO: Starting box [0, 0]
2020-07-03 13:03:13,778 INFO: Starting box [4091, 0]
...
This process seems very slow.  Is it necessary to convolve over a grid when there is only a single PSF?

**Comment by benw1** [comment]
From the example, it looks like the keyword to set is "psf_grid_size", but this does not work for me.  See below, where psf_grid_size is set to 3, but the code generates a 1x1 PSF grid.
>>> print(obs['psf_grid_size'])
3
>>> obm = ObservationModule(obs)
2020-07-03 16:43:15,947 INFO: Got offsets as [{'offset_id': 1, 'offset_centre': False, 'offset_ra': 0.5, 'offset_dec': 0.0, 'offset_pa': 27.0}]
2020-07-03 16:43:15,948 INFO: Adding observation with filter F158 and offset (0.500000,0.000000,27.000000)
2020-07-03 16:43:15,949 INFO: Added 1 observations
2020-07-03 16:43:15,958 INFO: WFI with 1 detectors. Central offset (0.0, 0.0, 0.0)
>>> obm.nextObservation()
2020-07-03 16:43:22,314 INFO: Initializing Observation 0 of 1
2020-07-03 16:43:22,317 INFO: Observation Filter is F158
2020-07-03 16:43:22,319 INFO: Observation (RA,DEC) = (0.000139,0.000000) with PA=27.000000
2020-07-03 16:43:22,319 INFO: Resetting
2020-07-03 16:43:25,328 INFO: Using local cache of JBT background data
2020-07-03 16:43:25,366 INFO: Creating Instrument with Configuration {'aperture': 'any', 'disperser': None, 'filter': 'f158', 'instrument': 'wfirstimager', 'mode': 'imaging'}
2020-07-03 16:43:25,512 INFO: Returning background 1.2581668704238262 for 'avg'
2020-07-03 16:43:25,512 INFO: Creating Detector SCA01 with (RA,DEC,PA) = (0.0001388888888888889,0.0,27.0)
2020-07-03 16:43:25,512 INFO: Creating Detector SCA01 with offset (0.0,0.0)
/nobackup/bwilli24/miniconda3/envs/stips/lib/python3.7/site-packages/pysynphot/units.py:851: RuntimeWarning: divide by zero encountered in log
  return -1.085736 * N.log(arg) + STZERO
2020-07-03 16:43:26,082 INFO: PSF choosing between 4096, 4096 and 215
2020-07-03 16:43:26,083 INFO: SCA01: Starting 1x1 PSF Grid creation at Fri Jul  3 16:43:26 2020

**Comment by york-stsci** [comment]
Okay, as far as I can see the above isn't related to PSF documentation, so I'm opening a new issue related to the PSF grid size keyword. That said, the convolution over a grid has nothing to do with PSF grids (except insofar as it makes PSF grids possible. This is done because whilst it is slow to convolve the PSF as as multiple overlapping segments, it's even slower if there isn't enough RAM available to hold both the full image and the full PSF in memory at the same time and, as a result, there ends up being a *lot* of paging in and out. In addition, on at least some systems, if python runs out of memory during a PSF convolution it dies with a segmentation fault instead of paging out, so splitting up the operation is the only way to avoid just having random crashes depending on the oversample and available RAM.

Note that it *is* possible to run the PSF convolution operations in parallel by supplying the parallel=True keyword argument to the ObservationModule. *NOTE* that this argument is supplied when creating the observation module itself, and having the keyword in your obs dictionary will not do anything.

If your system has plenty of RAM, it's possible to adjust the size of each individual convolution box (and thus the number of convolution boxes that will be used) by setting the convolve_size keyword to the ObservationModule (again, this is a keyword that goes to the module itself, and is *not* part of the obs dictionary). If you want to do the entire convolution as a single operation (*note* that doing this means that, even if you have a PSF grid, only the centre point of the grid will be used, since a single convolution is a single convolution), set the convolve_size to a power of 2 that is greater than

`(image_size)*oversample + psf_size/2 + psf_size`

Because of the way the operation is done, the convolution grid size at runtime is *always* a power of 2 so, if convolve_size is not a power of 2, then the next smallest power of 2 will be used.

**Comment by york-stsci** [comment]
This has effectively been superseded in version 2.0.0 by the new PSF application style, which generates a separate PSF for each source as the source is added.

**Comment by eunkyuh** [comment]
Fixed in Issue #92.
